src/react/utils.js

```diff
@@ -543,26 +543,41 @@ export function getComponentTypeFromRootValue(realm: Realm, value: Value): ECMAS
   }
 }
 
-// props should never have "ref" or "key" properties, as they're part of ReactElement
-// object instead. to ensure that we can give this hint, we create them and then
-// delete them, so their descriptor is left undefined. we use this knowledge later
-// to ensure that when dealing with creating ReactElements with partial config,
-// we don't have to bail out becuase "config" may or may not have "key" or/and "ref"
-export function deleteRefAndKeyFromProps(realm: Realm, props: ObjectValue | AbstractObjectValue): void {
-  setProperty(props, "ref", realm.intrinsics.undefined);
-  deleteProperty(props, "ref");
-  setProperty(props, "key", realm.intrinsics.undefined);
-  deleteProperty(props, "key");
-}
-
-export function objectHasNoPartialKeyAndRef(
-  realm: Realm,
-  object: ObjectValue | AbstractValue | AbstractObjectValue
-): boolean {
-  if (object instanceof AbstractValue) {
+export function flagPropsWithNoPartialKeyOrRef(realm: Realm, props: ObjectValue | AbstractObjectValue): void {
+  realm.react.propsWithNoPartialKeyOrRef.add(props);
+}
+
+export function hasNoPartialKeyOrRef(realm: Realm, props: ObjectValue | AbstractObjectValue): boolean {
+  if (realm.react.propsWithNoPartialKeyOrRef.has(props)) {
+    return true;
+  }
+  if (props instanceof ObjectValue && !props.isPartialObject()) {
+    realm.react.propsWithNoPartialKeyOrRef.add(props);
```

Comment:
It's counter-intuitive to me that a function with a "pure" name like `has*` performs this caching side effect.

How do we know it's never going to become partial? Some other code could call `makePartial()`.

Reply:
This is not a cache. It's adding it to the Set as it's not partial thus doesn't have partial properties.